### PR TITLE
add metrics flush thread

### PR DESCRIPTION
**Pull request by follitude**
PTAL @zhangjinpeng1987 @siddontang

### CODE REVIEW

**Comment by siddontang** [comment]
I think we should use another worker to flush the RocksDB metrics, not in raftstore thread now.
We can also let PD worker to calculate the store statistics, not in raftstore thread too.

File: src/raftstore/store/worker/flush_metrics.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+// limitations under the License.
+
+use rocksdb::DB;
+use std::sync::{Arc, RwLock};
+use super::super::engine_metrics::*;
+use std::thread::{self, JoinHandle, Builder};
+use std::io;
+use std::time;
+
+pub struct FlushMetrics {
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/worker/flush_metrics.rs]: MetricsFlusher

File: src/raftstore/store/worker/pd.rs [hunk 2]
```diff
@@ -157,7 +165,64 @@ impl<T: PdClient> Runner<T> {
...
+
+        let disk_cap = disk_stats.total_space();
+        let capacity = if store_info.capacity == 0 || disk_cap < store_info.capacity {
+            disk_cap
+        } else {
+            store_info.capacity
+        };
+        stats.set_capacity(capacity);
+
+        let mut used_size = flush_engine_properties_and_get_used_size(store_info.engine.clone());
```
Comment (siddontang) [inline src/raftstore/store/worker/pd.rs]: I prefer using get_used_size directly, you should flush engine properties to the flush thread.

**Comment by siddontang** [comment]
you must use 2 PRs, one is for flush metric, one is for store heartbeat.

**Comment by siddontang** [comment]
please
1. move the RocksDB metrics and flusher to util/rocksdb directory.
2. split flush_engine_properties_and_get_used_size and support get_used_size to let outer get the rocksdb size 
3. flush engine properties in the Flusher 
4. Use the Flusher in server, not in raftstore.

File: src/raftstore/store/worker/metrics_flush.rs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+                    for t in ENGINE_TICKER_TYPES {
+                        let v = db.get_statistics_ticker_count(*t);
+                        flush_engine_ticker_metrics(*t, v);
+                    }
+                    for t in ENGINE_HIST_TYPES {
+                        if let Some(v) = db.get_statistics_histogram(*t) {
+                            flush_engine_histogram_metrics(*t, v);
+                        }
+                    }
+                    thread::sleep(time::Duration::from_millis(1000));
```
Comment (siddontang) [inline src/raftstore/store/worker/metrics_flush.rs]: pass the interval value when creating the flusher. Btw, 1s is short.

you should use a channel to support recv_timeout, so you can stop the thread immediately, no need to wait a long time.

File: src/raftstore/store/worker/metrics_flush.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+// limitations under the License.
+
+use rocksdb::DB;
+use std::sync::{Arc, RwLock};
+use super::super::engine_metrics::*;
+use std::thread::{self, JoinHandle, Builder};
+use std::io;
+use std::time;
+
+pub struct MetricsFlusher {
```
Comment (siddontang) [inline src/raftstore/store/worker/metrics_flush.rs]: please add a simple test for the Flusher.

**Comment by siddontang** [comment]
any update?

File: src/server/config.rs [hunk 5]
```diff
@@ -32,6 +32,7 @@ const DEFAULT_GRPC_CONCURRENT_STREAM: usize = 1024;
 const DEFAULT_GRPC_RAFT_CONN_NUM: usize = 10;
 const DEFAULT_GRPC_STREAM_INITIAL_WINDOW_SIZE: u64 = 2 * 1024 * 1024;
 const DEFAULT_MESSAGES_PER_TICK: usize = 4096;
+const DEFAULT_FLUSH_METRICS_INTERVAL: u64 = 5000;
```
Comment (zhangjinpeng1987) [inline src/server/config.rs]: Use the original value 10 seconds.

File: src/util/rocksdb/metrics_flush.rs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+use util::rocksdb::engine_metrics::*;
+use std::thread::{self, JoinHandle, Builder};
+use std::io;
+use std::time;
+
+pub struct MetricsFlusher {
+    engine: Arc<DB>,
+    flag: Arc<RwLock<bool>>,
+    handle: Option<JoinHandle<()>>,
+    interval: Arc<RwLock<u64>>,
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flush.rs]: why use arc lock here?

File: src/server/config.rs [hunk 7]
```diff
@@ -53,7 +54,7 @@ pub struct Config {
     pub grpc_raft_conn_num: usize,
     pub grpc_stream_initial_window_size: ReadableSize,
     pub end_point_concurrency: usize,
-
+    pub flush_metrics_interval: u64,
```
Comment (siddontang) [inline src/server/config.rs]: we don't need this config, using default 10s is enough.

File: src/server/server.rs [hunk 8]
```diff
@@ -48,12 +50,15 @@ pub struct Server<T: RaftStoreRouter + 'static, S: StoreAddrResolver + 'static>
...
     snap_worker: Worker<SnapTask>,
+    metrics_flusher: Option<MetricsFlusher>,
 }
 
 impl<T: RaftStoreRouter, S: StoreAddrResolver + 'static> Server<T, S> {
+    #[allow(too_many_arguments)]
     pub fn new(cfg: &Config,
                region_split_size: usize,
                storage: Storage,
+               engine: Option<Arc<DB>>,
```
Comment (siddontang) [inline src/server/server.rs]: my fault, it is ugly to start rocksdb flusher in Server. maybe it is better to start the flusher directly in the tikv-server.rs after the storage starts.

File: src/util/rocksdb/metrics_flush.rs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+impl MetricsFlusher {
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flush.rs]: please add a test for the Flusher.

File: src/util/rocksdb/metrics_flush.rs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+    pub fn start(&mut self) -> Result<(), io::Error> {
+        let db = self.engine.clone();
+        let flag = self.flag.clone();
+        let interval = self.interval.clone();
+        let h = try!(Builder::new()
+            .name(thd_name!("flush metrics"))
+            .spawn(move || {
+                loop {
+                    if *(flag.read().unwrap()) {
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flush.rs]: if we stop the flusher, we may need to wait at lease interval time, this is slow when we close TiKV.

File: src/util/rocksdb/metrics_flush.rs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+                    thread::sleep(time::Duration::from_millis(*(interval.read().unwrap())));
+                }
+            }));
+
+        self.handle = Some(h);
+        Ok(())
+    }
+
+    pub fn stop(&mut self) -> Option<::std::thread::JoinHandle<()>> {
+        *(self.flag.write().unwrap()) = true;
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/metrics_flush.rs]: Is it means we maybe need to wait `interval`?

File: src/util/rocksdb/metrics_flush.rs (hunk 10 above)
Comment (siddontang) [inline src/util/rocksdb/metrics_flush.rs]: use channel and recv timeout instead.

File: src/util/rocksdb/metrics_flush.rs [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+            .unwrap();
+        if let Err(e) = metrics_flusher.start() {
+            error!("failed to start metrics flusher, error = {:?}", e);
+        }
+        server.start(&cfg).unwrap();
+
+        let rtime = Duration::from_millis(20000);
+        let now = Instant::now();
+
+        while now.elapsed() >= rtime {
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flush.rs]: sleep rtime

File: src/util/rocksdb/metrics_flush.rs [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+            .unwrap_or_else(|e| exit_with_err(e));
+        storage.start(&storage_cfg).unwrap();
+
+        let (snapshot_status_sender, _) = mpsc::channel();
+
+        let addr = Arc::new(Mutex::new(None));
+
+        let mut metrics_flusher = MetricsFlusher::new(engine.clone(), 10000);
+
+        let mut server = Server::new(&cfg,
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flush.rs]: you can remove server, no need here, just to test flusher.

File: src/util/rocksdb/metrics_flush.rs [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+        self.handle = Some(h);
+        Ok(())
+    }
+
+    pub fn stop(&mut self) -> Option<::std::thread::JoinHandle<()>> {
+        if self.handle.is_none() {
+            return None;
+        }
+        self.sender.as_ref().unwrap().send(true).unwrap();;
+        self.handle.take()
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flush.rs]: join the thread here directly

File: src/util/rocksdb/metrics_flush.rs [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+pub struct MetricsFlusher {
+    engine: Arc<DB>,
+    handle: Option<JoinHandle<()>>,
+    sender: Option<Sender<bool>>,
+    interval: u64,
+}
+
+impl MetricsFlusher {
+    pub fn new(engine: Arc<DB>, interval: u64) -> MetricsFlusher {
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flush.rs]: pass Duration here

File: src/bin/tikv-server.rs [hunk 16]
```diff
@@ -162,6 +163,9 @@ fn run_raft_server(pd_client: RpcClient, cfg: &TiKvConfig) {
     let snap_mgr = SnapManager::new(snap_path.as_path().to_str().unwrap().to_owned(),
                                     Some(store_sendch),
                                     cfg.raft_store.use_sst_file_snapshot);
+
+    let mut metrics_flusher = MetricsFlusher::new(engine.clone(), 10000);
```
Comment (siddontang) [inline src/bin/tikv-server.rs]: define a const value for 10000

File: src/util/rocksdb/metrics_flush.rs [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+        let storage_cfg = StorageConfig::default();
+        cfg.addr = "127.0.0.1:0".to_owned();
+        let (tx, _) = mpsc::channel();
+        let router = TestRaftStoreRouter::new(tx);
+        let engine = Arc::new(rocksdb::new_engine_opt(path.path().to_str().unwrap(),
+                                                      db_opt,
+                                                      vec![CFOptions::new(CF_DEFAULT, cf_opts)])
+            .unwrap());
+
+        let mut storage = create_raft_storage(router.clone(), engine.clone(), &storage_cfg)
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flush.rs]: remove storage

File: src/bin/tikv-server.rs [hunk 18]
```diff
@@ -188,12 +192,20 @@ fn run_raft_server(pd_client: RpcClient, cfg: &TiKvConfig) {
         panic!("failed to start storage, error = {:?}", e);
     }
 
+    if let Err(e) = metrics_flusher.start() {
```
Comment (zhangjinpeng1987) [inline src/bin/tikv-server.rs]: Can we construct the flusher just before start.

File: src/raftstore/store/store.rs [hunk 19]
```diff
@@ -192,7 +191,7 @@ impl<T, C> Store<T, C> {
         let mut s = Store {
             cfg: Rc::new(cfg),
             store: meta,
-            engine: engine,
+            engine: engine.clone(),
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: No need clone here.

File: src/util/rocksdb/engine_metrics.rs [hunk 20]
```diff
@@ -149,23 +269,37 @@ pub fn flush_engine_histogram_metrics(t: HistType, value: HistogramData) {
     }
 }
 
-pub fn flush_engine_properties_and_get_used_size(engine: Arc<DB>) -> u64 {
+pub fn get_used_size(engine: Arc<DB>) -> u64 {
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/engine_metrics.rs]: This function should placed in mod.rs.

File: src/util/rocksdb/metrics_flush.rs [hunk 21]
```diff
@@ -0,0 +1,160 @@
+// Copyright 2017 PingCAP, Inc.
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/metrics_flush.rs]: Rename this file to metrics_flusher.rs.

File: src/util/rocksdb/metrics_flush.rs [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+        let h = try!(Builder::new()
+            .name(thd_name!("flush metrics"))
+            .spawn(move || {
+                loop {
+                    match rx.recv_timeout(Duration::from_millis(interval)) {
+                        Ok(_) => {
+                            break;
+                        }
+                        Err(mpsc::RecvTimeoutError::Timeout) => {
+                            for t in ENGINE_TICKER_TYPES {
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/metrics_flush.rs]: Please move these code into an independent function.

File: src/server/server.rs [hunk 23]
```diff
@@ -145,7 +145,7 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver + 'static> Server<T, S> {
 }
 
 #[cfg(test)]
-mod tests {
+pub mod tests {
```
Comment (zhangjinpeng1987) [inline src/server/server.rs]: It is a bit of strange to make tests module public.

File: src/util/rocksdb/metrics_flush.rs [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+                                     snapshot_status_sender,
+                                     MockResolver { addr: addr.clone() },
+                                     SnapManager::new("", None, true))
+            .unwrap();
+        if let Err(e) = metrics_flusher.start() {
+            error!("failed to start metrics flusher, error = {:?}", e);
+        }
+        server.start(&cfg).unwrap();
+
+        let rtime = Duration::from_millis(20000);
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/metrics_flush.rs]: You can use a const var and redefine it make it more smaller in test module.

File: src/util/rocksdb/metrics_flush.rs [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+        let now = Instant::now();
+
+        while now.elapsed() >= rtime {
+            server.stop().unwrap();
+        }
+
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/metrics_flush.rs]: After flush metric, please check the content of Metrics.

File: src/util/rocksdb/metrics_flusher.rs [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+        let (tx, rx) = mpsc::channel();
+        let interval = self.interval;
+        self.sender = Some(tx);
+        let h = try!(Builder::new().name(thd_name!("flush metrics")).spawn(
+            move || {
+                let db = engine;
+                loop {
+                    match rx.recv_timeout(interval) {
+                        Ok(_) => {
+                            break;
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/metrics_flusher.rs]: Please add comments to explain when this would happen.

File: src/util/rocksdb/metrics_flusher.rs [hunk 27]
```diff
@@ -0,0 +1,134 @@
...
+        self.handle = Some(h);
+        Ok(())
+    }
+
+    pub fn stop(&mut self) {
+        let h = self.handle.take();
+        if h.is_none() {
+            return;
+        }
+        self.sender.as_ref().unwrap().send(true).unwrap();;
```
Comment (zhangjinpeng1987) [inline src/util/rocksdb/metrics_flusher.rs]: Please trim redundant `;` at the tail.

File: src/raftstore/store/store.rs [hunk 28]
```diff
@@ -1849,7 +1848,7 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         };
         stats.set_capacity(capacity);
 
-        let mut used_size = flush_engine_properties_and_get_used_size(self.engine.clone());
+        let mut used_size = get_used_size(self.engine.clone());
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/store.rs]: I think we should move `get_used_size` into pd-worker.

Comment (siddontang) [inline src/raftstore/store/store.rs]: we can do it in another PR, not here.

File: src/util/rocksdb/metrics_flusher.rs [hunk 29]
```diff
@@ -0,0 +1,134 @@
...
+            CFOptions::new(CF_RAFT, rocksdb::ColumnFamilyOptions::new()),
+            CFOptions::new(CF_LOCK, rocksdb::ColumnFamilyOptions::new()),
+            CFOptions::new(CF_WRITE, cf_opts),
+        ];
+        let engine = Arc::new(
+            rocksdb::new_engine_opt(path.path().to_str().unwrap(), db_opt, cfs_opts).unwrap(),
+        );
+        let mut metrics_flusher = MetricsFlusher::new(
+            engine.clone(),
+            Duration::from_millis(DEFAULT_FLUSHER_INTERVAL),
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flusher.rs]: 10s is long in the test, use 100ms is enough.

File: src/util/rocksdb/metrics_flusher.rs [hunk 30]
```diff
@@ -0,0 +1,134 @@
...
+        );
+
+        if let Err(e) = metrics_flusher.start() {
+            error!("failed to start metrics flusher, error = {:?}", e);
+        }
+
+        let rtime = Duration::from_millis(DEFAULT_RUN_TIME);
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flusher.rs]: 3 * interval is ok.

File: src/util/rocksdb/metrics_flusher.rs (hunk 29 above)
Comment (siddontang) [inline src/util/rocksdb/metrics_flusher.rs]: IMO, in test, you don't need to define a const var.

File: src/util/rocksdb/metrics_flusher.rs [hunk 31]
```diff
@@ -0,0 +1,134 @@
...
+    pub fn start(&mut self) -> Result<(), io::Error> {
+        let engine = self.engine.clone();
+        let (tx, rx) = mpsc::channel();
+        let interval = self.interval;
+        self.sender = Some(tx);
+        let h = try!(Builder::new().name(thd_name!("flush metrics")).spawn(
+            move || {
+                let db = engine;
+                loop {
+                    match rx.recv_timeout(interval) {
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flusher.rs]:
```
while if let Err(timeout) = rx.recv_timeout() {
}
```

File: src/util/rocksdb/metrics_flusher.rs (hunk 27 above)
Comment (siddontang) [inline src/util/rocksdb/metrics_flusher.rs]: no need send true, take to get the sender can drop is ok.

File: src/util/rocksdb/metrics_flusher.rs [hunk 32]
```diff
@@ -0,0 +1,121 @@
...
+            interval: interval,
+        }
+    }
+
+    pub fn start(&mut self) -> Result<(), io::Error> {
+        let engine = self.engine.clone();
+        let (tx, rx) = mpsc::channel();
+        let interval = self.interval;
+        self.sender = Some(tx);
+        let h = try!(Builder::new().name(thd_name!("flush metrics")).spawn(
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flusher.rs]: thread name should be "rocksb-metrics-flusher"

File: src/util/rocksdb/metrics_flusher.rs [hunk 33]
```diff
@@ -0,0 +1,121 @@
...
+            move || {
+                let db = engine;
+                while let Err(mpsc::RecvTimeoutError::Timeout) = rx.recv_timeout(interval) {
+                    flush_metrics(&db);
```
Comment (siddontang) [inline src/util/rocksdb/metrics_flusher.rs]: use engine directly

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

**Comment by zhangjinpeng1987** [comment]
LGTM
